state-chain/amm/src/lib.rs

```diff
@@ -519,3 +552,44 @@ impl<LiquidityProvider: Clone + Ord> PoolState<LiquidityProvider> {
 			range_orders::PoolState::<LiquidityProvider>::validate_fees(fee_hundredth_pips)
 	}
 }
+fn reduce_by_pool_fee(input: U256, fee_hundredth_pips: u32) -> U256 {
+	// This cannot overflow as we bound fee_hundredth_pips to <= ONE_IN_HUNDREDTH_PIPS/2
+	mul_div_floor(
+		input,
+		U256::from(ONE_IN_HUNDREDTH_PIPS - fee_hundredth_pips),
+		U256::from(ONE_IN_HUNDREDTH_PIPS),
+	)
+}
+
+fn grow_by_pool_fee(input: U256, fee_hundredth_pips: u32) -> U256 {
+	// This cannot overflow as we bound fee_hundredth_pips to <= ONE_IN_HUNDREDTH_PIPS/2
+	mul_div_floor(
+		input,
+		U256::from(ONE_IN_HUNDREDTH_PIPS),
+		U256::from(ONE_IN_HUNDREDTH_PIPS - fee_hundredth_pips),
+	)
+}
+
+/// Returns the ratio between price between prices with and without accounting for the pool fee
+/// as a fixed point number (same representation as Price).
+fn pool_fee_scaling_factor(side: Side, fee_hundredth_pips: u32) -> U256 {
+	use cf_amm_math::PRICE_FRACTIONAL_BITS;
+	match side {
+		Side::Buy => grow_by_pool_fee(U256::one() << PRICE_FRACTIONAL_BITS, fee_hundredth_pips),
+		Side::Sell => reduce_by_pool_fee(U256::one() << PRICE_FRACTIONAL_BITS, fee_hundredth_pips),
+	}
+}
+
+fn sqrt_price_adjusted_by_pool_fee<SD: common::SwapDirection>(
+	sqrt_price: SqrtPriceQ64F96,
+	fee_hundredth_pips: u32,
+) -> SqrtPriceQ64F96 {
+	use cf_amm_math::SQRT_PRICE_FRACTIONAL_BITS;
+
+	let fee_factor = pool_fee_scaling_factor(SD::INPUT_SIDE.sell_order(), fee_hundredth_pips);
+	// fee_factor isn't price, but we can use the same code as for the price
+	// because they use the same fixed point representation:
+	let sqrt_fee_factor = price_to_sqrt_price(fee_factor);
+
+	mul_div_floor(sqrt_price, sqrt_fee_factor, U256::one() << SQRT_PRICE_FRACTIONAL_BITS)
```

Comment:
Was it a conscious decision to do it this way (converting the factor) rather than the more obvious (converting the price)? 

```
let price = sqrt_price_to_price(sqrt_price);
let adjusted_price = match side {
	Side::Buy => grow_by_pool_fee(price, fee_hundredth_pips),
	Side::Sell => reduce_by_pool_fee(price, fee_hundredth_pips),
};
price_to_sqrt_price(adjusted_price)
```